tempesta_fw/client.c

```diff
@@ -28,122 +28,222 @@
 
-/* Total number of created clients. */
-static atomic64_t act_cli_n = ATOMIC64_INIT(0);
+unsigned long tfw_hash_str_len(const TfwStr *str, unsigned long str_len);
 
 /**
  * Called when a client socket is closed.
  */
 void
 tfw_client_put(TfwClient *cli)
 {
-	TFW_DBG2("put client %p, conn_users=%d\n",
-		 cli, atomic_read(&cli->conn_users));
+	TfwClientEntry *ent = (TfwClientEntry *)cli;
+
+	TFW_DBG2("put client %p, users=%d\n",
+		 cli, atomic_read(&ent->users));
 
-	if (!atomic_dec_and_test(&cli->conn_users))
+	if (!atomic_dec_and_test(&ent->users))
 		return;
 
-	spin_lock(cli->hb_lock);
+	spin_lock(&ent->lock);
 
-	if (atomic_read(&cli->conn_users)) {
-		spin_unlock(cli->hb_lock);
+	if (atomic_read(&ent->users)) {
+		spin_unlock(&ent->lock);
 		return;
 	}
 	BUG_ON(!list_empty(&cli->conn_list));
 
-	hlist_del(&cli->hentry);
+	ent->expires = tfw_current_timestamp() + client_cfg.lifetime;
 
-	spin_unlock(cli->hb_lock);
+	spin_unlock(&ent->lock);
 
-	TFW_DBG("free client: cli=%p\n", cli);
-	kmem_cache_free(cli_cache, cli);
-	atomic64_dec(&act_cli_n);
+	TFW_DBG("put client: cli=%p\n", cli);
 	TFW_DEC_STAT_BH(clnt.online);
 }
-EXPORT_SYMBOL(tfw_client_put);
 
-/**
- * Find a client corresponding to the @sk by IP address.
- * More advanced identification is possible based on User-Agent,
- * Cookie and other HTTP headers.
- * TODO (#488,#598,#1054?) actually clients can be relatively reliably
- * identified with TLS sessions (#1054) or HTTP sticky cookies.
- *
- * The returned TfwClient reference must be released via tfw_client_put()
- * when the @sk is closed.
- *
- * TODO #488 (#100): evict connections and/or clients and drop their accounting.
- * For now a client is freed immediately when the last of its connections is
- * closed, probably we should evict clients after some timeout to keep their
- * classifier statistic for following sessions...
- */
-TfwClient *
-tfw_client_obtain(struct sock *sk, void (*init)(TfwClient *))
-{
-	TfwClient *cli;
-	CliHashBucket *hb;
-	struct hlist_node *tmp;
-	unsigned long key;
-	TfwAddr addr;
+typedef struct {
+	TfwAddr *addr;
+	TfwAddr *cli_addr;
+	TfwStr	*user_agent;
+} TfwClientEqCtx;
 
-	ss_getpeername(sk, &addr);
-	key = hash_calc((const char *)&addr.sin6_addr,
-			sizeof(addr.sin6_addr));
+static struct in6_addr any_addr = IN6ADDR_ANY_INIT;
 
-	hb = &cli_hash[hash_min(key, CLI_HASH_BITS)];
+static bool
+tfw_client_addr_eq(TdbRec *rec, void (*init)(void *), void *data)
+{
+	TfwClientEntry *ent = (TfwClientEntry *)rec->data;
+	TfwClient *cli = &ent->cli;
+	TfwClientEqCtx *ctx = (TfwClientEqCtx *)data;
+	time_t curr_time = tfw_current_timestamp();
+	int users;
+
+	if (memcmp_fast(&cli->addr.sin6_addr, &ctx->addr->sin6_addr,
+			sizeof(cli->addr.sin6_addr)))
+	{
+		return false;
+	}
 
-	spin_lock(&hb->lock);
+	if ((!ctx->cli_addr && memcmp_fast(&ent->cli_addr.sin6_addr, &any_addr,
+					   sizeof(ent->cli_addr.sin6_addr))) ||
+	    (ctx->cli_addr && memcmp_fast(&ent->cli_addr.sin6_addr,
```

Comment:
There are already `if (cli_addr)` and `if (user_agent)` conditions in `tfw_client_obtain()`. And you can init `ctx->cli_addr` as `any_addr` and `ctx->user_agent` as empty string there to make conditions in this function simpler. But it's just a matter of taste.